ci/tests/richards.c

```diff
@@ -0,0 +1,432 @@
+
+            case S_WAIT:
+            case S_HOLD:
+            case S_HOLDPKT:
+            case S_HOLDWAIT:
+            case S_HOLDWAITPKT:
+                tcb = tcb->t_link;
+                break;
+
+            default:
+                return;
+        }
+    }
+}
+
+struct task *wait_task(void)
+{
+    tcb->t_state |= WAITBIT;
+    return (tcb);
+}
+
+struct task *holdself(void)
+{
+    ++holdcount;
+    tcb->t_state |= HOLDBIT;
+    return (tcb->t_link) ;
+}
+
+struct task *findtcb(int id)
+{
+    struct task *t = 0;
+
+    if (1<=id && id<=(long)tasktab[0])
+      t = tasktab[id];
+    if (t==0) 
+      printf("\nBad task id %d\n", id);
+    return(t);
+}
+
+struct task *release(int id)
+{
+    struct task *t;
+
+    t = findtcb(id);
+    if ( t==0 ) return (0);
+
+    t->t_state &= NOTHOLDBIT;
+    if ( t->t_pri > tcb->t_pri )
+        return (t);
+
+    return (tcb) ;
+}
+
+
+struct task *qpkt(struct packet *pkt)
+{
+    struct task *t;
+
+    t = findtcb(pkt->p_id);
+    if (t==0) return (t);
+
+    qpktcount++;
+
+    pkt->p_link = 0;
+    pkt->p_id = taskid;
+
+   if (t->t_wkq==0)
+    {
+        t->t_wkq = pkt;
+        t->t_state |= PKTBIT;
+        if (t->t_pri > tcb->t_pri) return (t);
+    }
+    else
+    {
+        append(pkt, (struct packet *)&(t->t_wkq));
+    }
+
+    return (tcb);
+}
+
+struct task *idlefn(struct packet *pkt)
+{
+    if ( --v2==0 ) return ( holdself() );
+
+    if ( (v1&1) == 0 )
+    {
+        v1 = ( v1>>1) & MAXINT;
+        return ( release(I_DEVA) );
+    }
+    else
+    {
+        v1 = ( (v1>>1) & MAXINT) ^ 0XD008;
+        return ( release(I_DEVB) );
+    }
+}
+
+struct task *workfn(struct packet *pkt)
+{
+    if ( pkt==0 ) { 
+        return ( wait_task() );
+    } else {
+        int i;
+
+        v1 = I_HANDLERA + I_HANDLERB - v1;
+        pkt->p_id = v1;
+
+        pkt->p_a1 = 0;
+        for (i=0; i<=BUFSIZE; i++)
+        {
+            v2++;
+            if ( v2 > 26 ) 
+              v2 = 1;
+            (pkt->p_a2)[i] = alphabet[v2];
+        }
+        return ( qpkt(pkt) );
+    }
+}
+
+struct task *handlerfn(struct packet *pkt)
+{
+  if ( pkt!=0) {
+    append(pkt, (struct packet *)(pkt->p_kind==K_WORK ? &v1 : &v2));
+  }
+
+  if ( v1!=0 ) {
+    int count;
+    struct packet *workpkt = (struct packet *)v1;  // dodgy cast
+    count = workpkt->p_a1;
+
+    if ( count > BUFSIZE ) {
+      v1 = (uintptr_t)(((struct packet *)v1)->p_link);  // dodgy cast
+      return ( qpkt(workpkt) );
+    }
+
+    if ( v2!=0 ) {
+      struct packet *devpkt;
+
+      devpkt = (struct packet *)v2;  // dodgy cast
+      v2 = (uintptr_t)(((struct packet *)v2)->p_link);
+      devpkt->p_a1 = workpkt->p_a2[count];
+      workpkt->p_a1 = count+1;
+      return( qpkt(devpkt) );
+    }
+  }
+  return wait_task();
+}
+
+struct task *devfn(struct packet *pkt)
+{
+    if ( pkt==0 )
+    {
+        if ( v1==0 ) 
+	    return ( wait_task() );
+        pkt = (struct packet *)v1;  // dodgy cast
```

Comment:
remove dodgy cast comment

Reply:
done in [2d9c363](https://github.com/capablevms/bdwgc/pull/18/commits/2d9c363a3cc31603584e349cb3b1ede700dea201)